codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -374,4 +374,762 @@ private String getSerializedMemberName(MemberShape memberShape) {
+        writer.openBlock("if !ok {", "}", () -> {
+            writer.write("return fmt.Errorf(\"expected $L to be of type $L, got %T instead\", st)"
+                    , shapeName, shapeSymbol.getName());
+        });
+
+        return CodegenUtils.generatePointerReferenceIfPointable(memberShape, "b");
+    }
+
+    // Generates deserializer for Byte, Short, Integer, Long, Big Integer member shape.
+    private String generateDocumentBindingIntegerMemberDeserializer(
+            GoWriter writer,
+            MemberShape memberShape
+    ) {
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("nt, err := val.(json.Number).Int64()");
+        writer.write("if err != nil { return err }");
+        switch (memberShape.getType()) {
+            case BYTE:
+                writer.write("st := byte(nt)");
+                break;
+            case SHORT:
+                writer.write("st := int16(nt)");
+            case INTEGER:
+                writer.write("st := int32(nt)");
+                break;
+            case LONG:
+                writer.write("st := nt");
+                break;
+            case BIG_INTEGER:
+                writer.addUseImports(GoDependency.BIG);
+                writer.write("st := big.NewInt(nt)");
+                break;
+        }
+
+        return CodegenUtils.generatePointerReferenceIfPointable(memberShape, "st");
+    }
+
+    // Generates deserializer for Float, Double, Big Decimal member shape.
+    private String generateDocumentBindingFloatMemberDeserializer(
+            GoWriter writer,
+            MemberShape memberShape
+    ) {
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("nt, err := val.(json.Number).Float64()");
+        writer.write("if err != nil { return err }");
+
+        switch (memberShape.getType()) {
+            case FLOAT:
+                writer.write("st := float32(nt)");
+                break;
+            case DOUBLE:
+                writer.write("st := nt");
+                break;
+            case BIG_DECIMAL:
+                writer.addUseImports(GoDependency.BIG);
+                writer.write("st := big.NewFloat(nt)");
+                break;
+        }
+
+        return CodegenUtils.generatePointerReferenceIfPointable(memberShape, "st");
+    }
+
+    // Generates deserializer for Timestamp member shape.
+    private String generateDocumentBindingTimestampMemberDeserializer(
+            GoWriter writer,
+            MemberShape memberShape
+    ) {
+        writer.write("val, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.write("nt, err := val.(json.Number).Int64()");
+        writer.write("if err != nil { return err }");
+
+        writer.addUseImports(GoDependency.TIME);
+        writer.write("ts := time.Unix(nt, 0).UTC()");
```

Comment:
Nit this should use the new `smithytime.ParseEpochSeconds(val.(string))` utility. Will also ensure that it handles fractional seconds too.

Reply:
Also if smithytime.ParseEpochSeconds is updated to return UTC the extra UTC call here won't be needed.